### PR TITLE
51124 - polling to check batch status

### PR DESCRIPTION
**Pull request by Vishal14583**
- FSS API client and polling configuration
- FSS Batch Status polling
- Changes for coding standards
- Detailed logging added for webjob
- Unit Tests added
- Functional Tests added

### CODE REVIEW

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment.UnitTests/Services/FssBatchServiceTests.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                .Should().Be("authFssTokenProvider");
+        }
+
+        [Test]
+        public async Task DoesCheckIfBatchCommitted_Returns_BatchStatus()
+        {
+            A.CallTo(() => _fakeFssApiClient.GetBatchStatusAsync(A<string>.Ignored, A<string>.Ignored))
+                .Returns(new HttpResponseMessage()
+                {
+                    StatusCode = System.Net.HttpStatusCode.OK,
```
Comment (shailen14599) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment.UnitTests/Services/FssBatchServiceTests.cs:69]: Can we have test for invalid batch (bad request)?

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.API.FunctionalTests/Helpers/GetBatchStatus.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
+﻿namespace UKHO.PeriodicOutputService.API.FunctionalTests.Helpers
+{
+    public class GetBatchStatus
+    {
+        static HttpClient httpClient = new HttpClient();
```
Comment (shailen14599) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.API.FunctionalTests/Helpers/GetBatchStatus.cs:5]:
```suggestion
        private static readonly HttpClient httpClient = new();
```
Can we do this change same as done in GetCatalogue.cs

**Review by shailen14599 — CHANGES_REQUESTED** [review]

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/AuthTokenProvider.cs [hunk 3]
```diff
@@ -1,30 +1,77 @@
...
+        private void AddAuthTokenToCache(string key, AccessTokenItem accessTokenItem)
+        {
+            var tokenExpiryMinutes = accessTokenItem.ExpiresIn.Subtract(DateTime.UtcNow).TotalMinutes;
+            var deductTokenExpiryMinutes = _essManagedIdentityConfiguration.Value.DeductTokenExpiryMinutes < tokenExpiryMinutes ? _essManagedIdentityConfiguration.Value.DeductTokenExpiryMinutes : 1;
+            var options = new DistributedCacheEntryOptions().SetSlidingExpiration(TimeSpan.FromMinutes(tokenExpiryMinutes - deductTokenExpiryMinutes));
+            options.SetAbsoluteExpiration(accessTokenItem.ExpiresIn);
+
+            lock (_lock)
+            {
+                _cache.SetString(key, JsonConvert.SerializeObject(accessTokenItem), options);
```
[the remark this answers is not in the document]
Comment (sumod13943) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/AuthTokenProvider.cs:69]: Added logs.

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/AuthFssTokenProvider.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+using System.Diagnostics.CodeAnalysis;
+using UKHO.PeriodicOutputService.Common.Configuration;
+
+namespace UKHO.PeriodicOutputService.Common.Helpers
+{
+    [ExcludeFromCodeCoverage] ////Excluded from code coverage as it has AD interaction
+    public class AuthFssTokenProvider : AuthTokenProvider, IAuthFssTokenProvider
+    {
+        public AuthFssTokenProvider(IOptions<EssManagedIdentityConfiguration> essManagedIdentityConfiguration, IDistributedCache _cache, ILogger<AuthFssTokenProvider> logger) :
+            base(essManagedIdentityConfiguration, _cache)
```
[the remark this answers is not in the document]
Comment (sumod13943) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/AuthFssTokenProvider.cs]: Parameter logger added. Resolved.

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/AuthEssTokenProvider.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+using System.Diagnostics.CodeAnalysis;
+using UKHO.PeriodicOutputService.Common.Configuration;
+
+namespace UKHO.PeriodicOutputService.Common.Helpers
+{
+    [ExcludeFromCodeCoverage] ////Excluded from code coverage as it has AD interaction
+    public class AuthEssTokenProvider : AuthTokenProvider, IAuthEssTokenProvider
+    {
+        public AuthEssTokenProvider(IOptions<EssManagedIdentityConfiguration> essManagedIdentityConfiguration, IDistributedCache _cache, ILogger<AuthEssTokenProvider> logger) :
+            base(essManagedIdentityConfiguration, _cache)
```
[the remark this answers is not in the document]
Comment (sumod13943) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/AuthEssTokenProvider.cs]: Parameter logger added. Resolved.

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/CommonHelper.cs [hunk 6]
```diff
@@ -12,5 +14,10 @@ public static string ExtractAccessToken(string response)
         {
             return response.Split(",")[0].Split(":")[1].Remove(0, 1).Replace("\"", "");
         }
+
+        public static string ExtractBatchId(string url)
+        {
+            return new UriBuilder(url).Uri.Segments.FirstOrDefault(d => Guid.TryParse(d.Replace("/", ""), out Guid _));
+        }
```
[the remark this answers is not in the document]
Comment (sumod13943) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/CommonHelper.cs]: Used in FSSBatchService.

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/Extensions.cs [hunk 7]
```diff
@@ -0,0 +1,17 @@
+﻿using System.Net.Http.Headers;
+
+namespace UKHO.PeriodicOutputService.Common
+{
+    public static class Extensions
```
[the remark this answers is not in the document]
Comment (sumod13943) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/Extensions.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/AccessTokenItem.cs [hunk 8]
```diff
@@ -0,0 +1,11 @@
+﻿using System.Diagnostics.CodeAnalysis;
+
+namespace UKHO.PeriodicOutputService.Common.Models
+{
+    [ExcludeFromCodeCoverage]
+    public class AccessTokenItem
+    {
+        public string AccessToken { get; set; } = string.Empty;
```
[the remark this answers is not in the document]
Comment (sumod13943) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/AccessTokenItem.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment.UnitTests/Services/FssBatchServiceTests.cs [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+                                                                                  BatchStatusPollingCutoffTime = "1",
+                                                                                  BatchStatusPollingDelayTime = "20000"});
+
+            _fakeLogger = A.Fake<ILogger<FssBatchService>>();
+
+            _fakeFssApiClient = A.Fake<IFssApiClient>();
+
+            _fakeAuthFssTokenProvider = A.Fake<IAuthFssTokenProvider>();
+
+            _fakeBatchService = new FssBatchService(_fakeLogger, _fakeFssApiConfiguration, _fakeFssApiClient, _fakeAuthFssTokenProvider);
```
[the remark this answers is not in the document]
Comment (sumod13943) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment.UnitTests/Services/FssBatchServiceTests.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment.UnitTests/Services/FssBatchServiceTests.cs (hunk 1 above)
Comment (sumod13943) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment.UnitTests/Services/FssBatchServiceTests.cs:70]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.API.FunctionalTests/Helpers/GetBatchStatus.cs (hunk 2 above)
Comment (sumod13943) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.API.FunctionalTests/Helpers/GetBatchStatus.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/Extensions.cs [hunk 10]
```diff
@@ -0,0 +1,19 @@
+﻿using System.Diagnostics.CodeAnalysis;
+using System.Net.Http.Headers;
+
+namespace UKHO.PeriodicOutputService.Common
+{
+    [ExcludeFromCodeCoverage]
+    public static class HttpRequestMessageExtensions
```
Comment (shailen14599) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/Extensions.cs:7]: Please rename the file name similar to class name.

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs [hunk 11]
```diff
@@ -0,0 +1,112 @@
+﻿using Microsoft.Extensions.Logging;
+
+namespace UKHO.PeriodicOutputService.Common.Logging
+{
+    public enum EventIds
+    {
+        /// <summary>
+        /// 809000 - Periodic Output Service webjob request started 
+        /// </summary
```
Comment (shailen14599) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs:9]: Please correct syntax wherever required.
```suggestion
        /// </summary>
```

**Review by shailen14599 — CHANGES_REQUESTED** [review]

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/Extensions.cs (hunk 10 above)
Comment (Vishal14583) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/Extensions.cs]: renamed

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs (hunk 11 above)
Comment (Vishal14583) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs]: Corrected

**Review by shailen14599 — APPROVED** [review]
LGTM

**Review by shailen14599 — APPROVED** [review]
LGTM

**Review by ajasonca — CHANGES_REQUESTED** [review]
Some changes around the ESS/FSS Auth Tokens to look at.

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/AuthTokenProvider.cs [hunk 12]
```diff
@@ -1,30 +1,84 @@
...
+using Newtonsoft.Json;
+using UKHO.PeriodicOutputService.Common.Configuration;
+using UKHO.PeriodicOutputService.Common.Models;
+using UKHO.PeriodicOutputService.Common.Logging;
 
 namespace UKHO.PeriodicOutputService.Common.Helpers
 {
-    public class AuthTokenProvider : IAuthTokenProvider
+    [ExcludeFromCodeCoverage] ////Excluded from code coverage as it has ADD interaction
+    public class AuthTokenProvider
```
[the remark this answers is not in the document]
Comment (Vishal14583) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/AuthTokenProvider.cs]: Resovled

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/AuthFssTokenProvider.cs [hunk 13]
```diff
@@ -0,0 +1,17 @@
+﻿using Microsoft.Extensions.Caching.Distributed;
```
[the remark this answers is not in the document]
Comment (Vishal14583) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/AuthFssTokenProvider.cs]: Removed

**Pull request merged**